cabal.project

```diff
@@ -1,3 +1,13 @@
-packages: ./base-compat
-          ./base-compat-batteries
-          ./check
+packages:
+  base-compat
+  base-compat-batteries
+  check
+
+package base-compat
+  ghc-options: -Werror
```

Comment:
This will enforce `-Werror` on CI, which is good, but it also means that local development will also fail if you have any warnings. I have mixed feelings about this, since it's pretty common to try out intermediate states during development that don't pass `-Werror` until you figure out what the final shape of a commit should be. Is there a way to only enforce `-Werror` on CI, but not during local development?

Reply:
Yes, ideally you only want it only on request (e.g. CI or before pushing).  This is possible with `stack` and `cabal v1-*`, but not with the `cabal v2-*` set of commands.

To make it work with `cabal v2-*` you would need to modify `cabal.project.local` after running `cabal configure --enable-tests --enable-benchmarks`.  However, there is no easy generic way to do that (you would need to look at all the cabal files that are referenced in `cabal.project` and extract the package names from these; this is why it's currently not part of `sol/run-haskell-tests@v1`).

For this repository specifically, the following should work:
```yaml
- run: cabal update
- run: cabal configure --enable-tests --enable-benchmarks # NOTE: --enable-tests is necessary due to https://github.com/haskell/cabal/issues/5079
- run: echo -e "package base-compat\n  ghc-options: -Werror\npackage base-compat-batteries\n  ghc-options: -Werror\npackage type-check\n  ghc-options: -Werror" >> cabal.project.local
- run: cabal build all
- run: cabal test all --test-show-details=direct
  env:
    HSPEC_OPTIONS: --color
```
(instead of `uses: sol/run-haskell-tests@v1`)

Reply:
Indeed, `haskell-ci` itself goes through each `.cabal` file and generates a `cabal.project.local` file based on that. But I agree that that's quite sophisticated, so I would also be content with just hardcoding `echo -e "package base-compat\n  ghc-options: -Werror\npackage base-compat-batteries\n  ghc-options: -Werror\npackage type-check\n  ghc-options: -Werror" >> cabal.project.local` as you suggest.